<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools for dealing with GWAS summary data in VCF format • gwasvcf</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Tools for dealing with GWAS summary data in VCF format">
<meta property="og:description" content="Tools for dealing with GWAS summary data in VCF format. Includes reading, querying, writing, as well as helper functions such as LD proxy searches.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">gwasvcf</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/guide.html">Reading, querying and writing GWAS summary data in VCF format</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrcieu/gwasvcf/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="reading-querying-and-writing-gwas-summary-data-in-vcf-format" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#reading-querying-and-writing-gwas-summary-data-in-vcf-format" class="anchor"></a>Reading, querying and writing GWAS summary data in VCF format</h1></div>
<!-- badges: start -->

<p>Complete GWAS summary datasets are now abundant. A large repository of curated, harmonised and QC’d datasets is available in the <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a>. They can be queried via the <a href="http://gwasapi.mrcieu.ac.uk/docs/">API</a> directly, or through the <a href="https://github.com/mrcieu/ieugwasr">ieugwasr</a> R package, or the <a href="https://github.com/mrcieu/ieugwaspy">ieugwaspy</a> python package. However, for faster querying that can be used in a HPC environment, accessing the data directly and not through cloud systems is advantageous.</p>
<p>We developed a format for storing and harmonising GWAS summary data known as <a href="https://github.com/MRCIEU/gwas_vcf_spec">GWAS VCF format</a>. All the data in the <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a> is available for download in this format. This R package provides fast and convenient functions for querying and creating GWAS summary data in GWAS VCF format. This package includes:</p>
<ul>
<li>a wrapper around the <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html">bioconductor/VariantAnnotation</a> package, providing functions tailored to GWAS VCF for reading, querying, creating and writing GWAS VCF format files</li>
<li>some LD related functions such as using a reference panel to extract proxies, create LD matrices and perform LD clumping</li>
<li>functions for harmonising a dataset against the reference genome and creating GWAS VCF files.</li>
</ul>
<p>See also the <a href="https://github.com/MRCIEU/gwasglue">gwasglue</a> R package for methods to connect the VCF data to Mendelian randomization, colocalisation, fine mapping etc.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"mrcieu/gwasvcf"</span><span class="op">)</span></code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>See vignettes here: <a href="https://mrcieu.github.io/gwasvcf">https://mrcieu.github.io/gwasvcf</a>.</p>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>If using GWAS-VCF files please reference the studies that you use and the following paper:</p>
<p><strong>The variant call format provides efficient and robust storage of GWAS summary statistics.</strong> Matthew Lyon, Shea J Andrews, Ben Elsworth, Tom R Gaunt, Gibran Hemani, Edoardo Marcora. bioRxiv 2020.05.29.115824; doi: <a href="https://doi.org/10.1101/2020.05.29.115824" class="uri">https://doi.org/10.1101/2020.05.29.115824</a></p>
</div>
<div id="reference-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-datasets" class="anchor"></a>Reference datasets</h2>
<p>Example GWAS VCF (GIANT 2010 BMI):</p>
<ul>
<li><a href="http://fileserve.mrcieu.ac.uk/vcf/IEU-a-2.vcf.gz">http://fileserve.mrcieu.ac.uk/vcf/IEU-a-2.vcf.gz</a></li>
<li><a href="http://fileserve.mrcieu.ac.uk/vcf/IEU-a-2.vcf.gz.tbi">http://fileserve.mrcieu.ac.uk/vcf/IEU-a-2.vcf.gz.tbi</a></li>
</ul>
<p>1000 genomes reference panels for LD for each superpopulation - used by default in OpenGWAS:</p>
<ul>
<li><a href="http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz">http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz</a></li>
</ul>
<p>1000 genomes European reference panel for LD (legacy):</p>
<ul>
<li><a href="http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz">http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz</a></li>
</ul>
<p>1000 genomes vcf harmonised against human genome reference:</p>
<ul>
<li><a href="http://fileserve.mrcieu.ac.uk/vcf/1kg_v3_nomult.vcf.gz">http://fileserve.mrcieu.ac.uk/vcf/1kg_v3_nomult.vcf.gz</a></li>
<li><a href="http://fileserve.mrcieu.ac.uk/vcf/1kg_v3_nomult.vcf.gz.tbi">http://fileserve.mrcieu.ac.uk/vcf/1kg_v3_nomult.vcf.gz.tbi</a></li>
</ul>
<hr>
<div id="notes" class="section level3">
<h3 class="hasAnchor">
<a href="#notes" class="anchor"></a>Notes</h3>
<div id="example-data" class="section level4">
<h4 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h4>
<p>data.vcf.gz and data.vcf.gz.tbi are the first few rows of the Speliotes 2010 BMI GWAS</p>
<p>The eur.bed/bim/fam files are the same range as data.vcf.gz, from here <a href="http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz" class="uri">http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz</a></p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/mrcieu/gwasvcf/">https://​github.com/​mrcieu/​gwasvcf/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mrcieu/gwasvcf/issues">https://​github.com/​mrcieu/​gwasvcf/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Gibran Hemani <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-0920-1055" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://codecov.io/github/mrcieu/gwasvcf"><img src="https://codecov.io/github/mrcieu/gwasvcf/branch/master/graphs/badge.svg" alt="codecov"></a></li>
<li><a href="https://github.com/MRCIEU/gwasvcf/actions"><img src="https://github.com/MRCIEU/gwasvcf/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
