<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Query data from vcf file — query_gwas • gwasvcf</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Query data from vcf file — query_gwas"><meta property="og:description" content="Read in GWAS summary data with filters on datasets (if multiple datasets per file) and/or chromosome/position, rsids or pvalues. Chooses most optimal choice for the detected operating system. Typically chrompos searches are the fastest. On Windows, rsid or pvalue filters from a file will be slow."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gwasvcf</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/guide.html">Reading, querying and writing GWAS summary data in VCF format</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mrcieu/gwasvcf/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Query data from vcf file</h1>
    <small class="dont-index">Source: <a href="https://github.com/mrcieu/gwasvcf/blob/HEAD/R/query.r" class="external-link"><code>R/query.r</code></a></small>
    <div class="hidden name"><code>query_gwas.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Read in GWAS summary data with filters on datasets (if multiple datasets per file) and/or chromosome/position, rsids or pvalues. Chooses most optimal choice for the detected operating system. Typically chrompos searches are the fastest. On Windows, rsid or pvalue filters from a file will be slow.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">query_gwas</span><span class="op">(</span></span>
<span>  <span class="va">vcf</span>,</span>
<span>  chrompos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rsid <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pval <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rsidx <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pvali <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  build <span class="op">=</span> <span class="st">"GRCh37"</span>,</span>
<span>  os <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.info.html" class="external-link">Sys.info</a></span><span class="op">(</span><span class="op">)</span><span class="op">[[</span><span class="st">"sysname"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  proxies <span class="op">=</span> <span class="st">"no"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dbfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tag_kb <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  tag_nsnp <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  tag_r2 <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>vcf</dt>
<dd><p>Path or URL to GWAS-VCF file or VCF object e.g. output from VariantAnnotation::readVcf or gwasvcftools::query_vcf</p></dd>


<dt>chrompos</dt>
<dd><p>Either vector of chromosome and position ranges e.g. "1:1000" or "1:1000-2000", or data frame with columns `chrom`, `start`, `end`.</p></dd>


<dt>rsid</dt>
<dd><p>Vector of rsids</p></dd>


<dt>pval</dt>
<dd><p>P-value threshold (NOT -log10)</p></dd>


<dt>id</dt>
<dd><p>If multiple GWAS datasets in the vcf file, the name (sample ID) from which to perform the filter</p></dd>


<dt>rsidx</dt>
<dd><p>Path to rsidx index file</p></dd>


<dt>pvali</dt>
<dd><p>Path to pval index file</p></dd>


<dt>build</dt>
<dd><p>="GRCh37" Build of vcffile</p></dd>


<dt>os</dt>
<dd><p>The operating system. Default is as detected. Determines the method used to perform query</p></dd>


<dt>proxies</dt>
<dd><p>="no" If SNPs are absent then look for proxies (yes) or not (no). Can also mask all target SNPs and only return proxies (only), for testing purposes. Currently only possible if querying rsid.</p></dd>


<dt>bfile</dt>
<dd><p>=path to plink bed/bim/fam ld reference panel</p></dd>


<dt>dbfile</dt>
<dd><p>=path to sqlite tag snp database</p></dd>


<dt>tag_kb</dt>
<dd><p>=5000 Proxy parameter</p></dd>


<dt>tag_nsnp</dt>
<dd><p>=5000 Proxy parameter</p></dd>


<dt>tag_r2</dt>
<dd><p>=0.6 Proxy parameter</p></dd>


<dt>threads</dt>
<dd><p>=1 NUmber of threads</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>vcf object</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

